styx-common/src/main/java/com/spotify/styx/storage/BigtableStorage.java

```diff
@@ -107,29 +109,46 @@ void writeEvent(SequenceEvent sequenceEvent) throws IOException {
     });
   }
 
-  List<WorkflowInstanceExecutionData> executionData(WorkflowId workflowId)
+  List<WorkflowInstanceExecutionData> executionData(WorkflowId workflowId, String offset, int limit)
       throws IOException {
     final Table eventsTable = connection.getTable(EVENTS_TABLE_NAME);
 
     final Scan scan = new Scan()
         .setRowPrefixFilter(Bytes.toBytes(workflowId.toKey() + '#'))
         .setFilter(new FirstKeyOnlyFilter());
 
-    final Set<WorkflowInstance> workflowInstancesSet = Sets.newHashSet();
-    for (Result result : eventsTable.getScanner(scan)) {
-      final String key = new String(result.getRow());
-      final int lastHash = key.lastIndexOf('#');
-      final WorkflowInstance wfi = WorkflowInstance.parseKey(key.substring(0, lastHash));
-      workflowInstancesSet.add(wfi);
+    if (!Strings.isNullOrEmpty(offset)) {
+      final WorkflowInstance offsetInstance = WorkflowInstance.create(workflowId, offset);
+      scan.setStartRow(Bytes.toBytes(offsetInstance.toKey() + '#'));
     }
 
-    final List<WorkflowInstanceExecutionData> workflowInstanceDataList = Lists.newArrayList();
-    for (WorkflowInstance workflowInstance : workflowInstancesSet) {
-      workflowInstanceDataList.add(executionData(workflowInstance));
+    final Set<WorkflowInstance> workflowInstancesSet = Sets.newHashSet();
+    try (ResultScanner scanner = eventsTable.getScanner(scan)) {
+      Result result = scanner.next();
+      while (result != null) {
+        final String key = new String(result.getRow());
+        final int lastHash = key.lastIndexOf('#');
+
+        final WorkflowInstance wfi = WorkflowInstance.parseKey(key.substring(0, lastHash));
+        workflowInstancesSet.add(wfi);
+        if (workflowInstancesSet.size() == limit) {
+          break;
+        }
+
+        result = scanner.next();
+      }
     }
-    workflowInstanceDataList.sort(WorkflowInstanceExecutionData.COMPARATOR);
 
-    return workflowInstanceDataList;
+    return workflowInstancesSet.parallelStream()
+        .map(workflowInstance -> {
+          try {
+            return executionData(workflowInstance);
```

Comment:
This is only partly related since it wasn't changed in this commit, but this function will call `readEvents` which in turns reads from the same table we just read from to get this list. Is there a point to that? Why not just use that data in the first scan?